Docs:Modify the flink version from 0.11.1 to 0.14.4 and fix several errors. (#4665)
apache/iceberg: docs/flink/flink-getting-started.md

## Preparation when using Flink SQL Client

> To create iceberg table in flink, we recommend to use [Flink SQL Client](https://nightlies.apache.org/flink/flink-docs-release-1.15/docs/dev/table/sqlclient/) because it's easier for users to understand the concepts.

Reviewer: Why change this link to nighlies.apache.org ?  I see the ci.apache.org will redirect to the target page correctly ?

Reviewer 2: +1 to not using nightlies.

Like Flink, we template our docs based on the Iceberg version now so it might be possible to do the same.

Otherwise, it used to be the case that the “latest” version (latest Spark version supported etc) could be automatically updated when the site was deployed.

So we might see if we can link to the correct version of the Flink docs, but otherwise the stable latest is better than nightlies imo.

Author: updated

Author: @openinx updated.

## Preparation when using Flink SQL Client

To create iceberg table in flink, we recommend to use [Flink SQL Client](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/dev/table/sqlclient/) because it's easier for users to understand the concepts.

Step.1 Downloading the flink 1.14.x binary package from the apache flink [download page](https://flink.apache.org/downloads.html). We now use scala 2.12 to archive the apache iceberg-flink-runtime jar, so it's recommended to use flink 1.14 bundled with scala 2.12.

```bash

> FLINK_VERSION=1.14.0

Reviewer 2: Nit: Maybe consider making this 1.14.4 (or whatever the latest is)? That's what they'll get if they go to the flink downloads page, but it's a minor detail either way.

Reviewer 2: Edit - I don't think this is necessary given that we refer to 1.14.0 in much of the rest of the document.

Author: Yes, i have change it to 1.15.0

```bash
# HADOOP_HOME is your hadoop root directory after unpack the binary package.
export HADOOP_CLASSPATH=`$HADOOP_HOME/bin/hadoop classpath`

./bin/sql-client.sh embedded -j <flink-runtime-directory>/iceberg-flink-runtime-1.15-xxx.jar shell
```

By default, iceberg has included hadoop jars for hadoop catalog. If we want to use hive catalog, we will need to load the hive jars when opening the flink sql client. Fortunately, apache flink has provided a [bundled hive jar](https://repo.maven.apache.org/maven2/org/apache/flink/flink-sql-connector-hive-2.3.6_2.12/1.15.0/flink-sql-connector-hive-2.3.6_2.12-1.15.0.jar) for sql client. So we could open the sql client
as the following:

```bash
# HADOOP_HOME is your hadoop root directory after unpack the binary package.
export HADOOP_CLASSPATH=`$HADOOP_HOME/bin/hadoop classpath`

# download Iceberg dependency

> ICEBERG_VERSION=0.13.1

Reviewer: The iceberg 0.13.1 does not support flink 1.15 currently.  Maybe it's better to use flink 1.14 in this example page.

Author: @openinx updated.

MAVEN_URL=https://repo.maven.apache.org/maven2
ICEBERG_MAVEN_URL=${MAVEN_URL}/org/apache/iceberg

> ICEBERG_FLINK_VERSION=1.15

Reviewer: This script won't work because  iceberg 0.13.1 does not support flink 1.15 .

Author: @openinx 
Yes, I used Flink 1.14.3 at the beginning, I can be sure that all versions are completely correct, but after I modified the Merge, no one helped me merge it for a long time. I thought it was version upgrade, so I changed it to 1.15.0.

Author: I have modified it to flink 1.14.4

Author: @openinx updated.

We've created a separate `flink-runtime` module in iceberg project to generate a bundled jar, which could be loaded by flink SQL client directly.

If we want to build the `flink-runtime` bundled jar manually for a specific Flink version, such as v1.14, please just build the `iceberg` project and it will generate the jar under `<iceberg-root-dir>/flink/v1.14/flink-runtime/build/libs`. Of course, we could also download the `flink-runtime-1.14` jar from the [Apache official repository](https://repo.maven.apache.org/maven2/org/apache/iceberg/iceberg-flink-runtime-1.14/).

```bash
# HADOOP_HOME is your hadoop root directory after unpack the binary package.
export HADOOP_CLASSPATH=`$HADOOP_HOME/bin/hadoop classpath`

./bin/sql-client.sh embedded -j <flink-runtime-directory>/iceberg-flink-runtime-1.14-xxx.jar shell
```

> By default, iceberg has included hadoop jars for hadoop catalog. If we want to use hive catalog, we will need to load the hive jars when opening the flink sql client. Fortunately, apache flink has provided a [bundled hive jar](https://repo.maven.apache.org/maven2/org/apache/flink/flink-sql-connector-hive-2.3.6_2.11/1.14.0/flink-sql-connector-hive-2.3.6_2.11-1.14.0.jar) for sql client. So we could open the sql client

Reviewer 2: Nit: I think this should use the Scala 2.12 link, which is `https://repo.maven.apache.org/maven2/org/apache/flink/flink-sql-connector-hive-2.3.6_2.12/1.14.0/flink-sql-connector-hive-2.3.6_2.12-1.14.0.jar`

Author: @kbendick I have change it,thank you.

We've created a separate `flink-runtime` module in iceberg project to generate a bundled jar, which could be loaded by flink SQL client directly.

> If we want to build the `flink-runtime` bundled jar manually, please just build the `iceberg` project and it will generate the jar under `<iceberg-root-dir>/flink/vx.xx/flink-runtime/build/libs`. Of course, we could also download the `flink-runtime` jar from the [apache official repository](https://repo.maven.apache.org/maven2/org/apache/iceberg/iceberg-flink-runtime/).

Reviewer 2: If we're updating this, the link https://repo.maven.apache.org/maven2/org/apache/iceberg/iceberg-flink-runtime/ is no longer correct.

The Flink 1.14 link would be https://repo.maven.apache.org/maven2/org/apache/iceberg/iceberg-flink-runtime-1.14.

Maybe it would be better to specifically mention the fact that there's a runtime jar per Flink version?

Something like
```
If we want to build the `flink-runtime` bundled jar manually for a specific Flink version, such as v1.14, please just build the `iceberg` project and it will generate the jar under `<iceberg-root-dir>/flink/v1.14/flink-runtime/build/libs`. Of course, we could also download the `flink-runtime-1.14` jar from the [apache official repository](If we want to build the `flink-runtime` bundled jar manually, please just build the `iceberg` project and it will generate the jar under `<iceberg-root-dir>/flink/vx.xx/flink-runtime/build/libs`. Of course, we could also download the `flink-runtime` jar from the [apache official repository](https://repo.maven.apache.org/maven2/org/apache/iceberg/iceberg-flink-runtime-1.14/)
```

I don't love the phrasing or the usage of pronouns (we) in this, and the instructions should probably mention building with the flinkVersions flag specified, but given that the current link is a valid link (but only for older code and likely won't work properly), it shouldn't be used.

                      "'my-additional-catalog-config'='my-value')")
```

For more details, please refer to the [Python Table API](https://ci.apache.org/projects/flink/flink-docs-release-1.14/docs/dev/python/table/intro_to_table_api/).

## Creating catalogs and using catalogs.

> Flink 1.14 support to create catalogs by using flink sql.

Reviewer 2: Maybe the version number can be dropped here? I think at one point in time Flink didn't support creating catalogs via SQL, only the YAML files.

So this is simply one less thing to keep updating if we remove the version number from it.

Author: @kbendick Thank you for your review, now I have revised all the questions, please check thank you.

Author: @kbendick  I want to upgrade to flink 1.15.0, but I find some resources are not available yet, so I still use 1.14.4.

Author: @kbendick Do i have something not modify ? please tell me : )

);
```

> Table create commands support the most commonly used [flink create clauses](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/dev/table/sql/create/) now, including: 

Reviewer 3: ```suggestion
Table create commands support the most commonly used [flink create clauses](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/dev/table/sql/create/), including: 
```
(nit)

Author: @samredai Thank you for your reply. I have changed the all the advise.

> Table create commands support the most commonly used [flink create clauses](https://ci.apache.org/projects/flink/flink-docs-release-1.11/dev/table/sql/create.html#create-table) now, including: 

Reviewer 2: Maybe consider updating this link from 1.11 to 1.14 as well? https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/dev/table/sql/create/ is the new link.

Author: I have changed the flink version from 1.11 to 1.14，please review this change, thank you.

CREATE TABLE  `hive_catalog`.`default`.`sample_like` LIKE `hive_catalog`.`default`.`sample`;
```

For more details, refer to the [Flink `CREATE TABLE` documentation](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/dev/table/sql/create/).

### `ALTER TABLE`

> Iceberg only support altering table properties in flink 1.14 now.

Reviewer 3: ```suggestion
Iceberg only supports altering table properties in flink 1.14.
```

## Writing with SQL

> Iceberg support both `INSERT INTO` and `INSERT OVERWRITE` in flink 1.14 now.

Reviewer 3: ```suggestion
Iceberg supports both `INSERT INTO` and `INSERT OVERWRITE` in flink 1.14.
```
